examples/Heat/mms.py

```diff
@@ -0,0 +1,200 @@
+
+
+# # # === --- finite element forms --- === # # #
+
+
+# asQ assumes that the mass form is linear so here
+# q is a TrialFunction and phi is a TestFunction
+def form_mass(q, phi):
+    return phi*q*fd.dx
+
+
+# q is a Function and phi is a TestFunction
+def form_function(q, phi, t):
+    return time_coef(t)*fd.inner(fd.grad(q), fd.grad(phi))*fd.dx - fd.inner(Rhs(t), phi)*fd.dx
+
+
+# # # === --- PETSc solver parameters --- === # # #
+
+
+# The PETSc solver parameters used to solve the
+# blocks in step (b) of inverting the ParaDiag matrix.
+block_parameters = {
+    'ksp_type': 'gmres',
+    'pc_type': 'bjacobi',
+}
+
+# The PETSc solver parameters for solving the all-at-once system.
+# The python preconditioner 'asQ.DiagFFTPC' applies the ParaDiag matrix.
+#
+# The equation is linear so we can either:
+# a) Solve it in one shot using a preconditioned Krylov method:
+#    P^{-1}Au = P^{-1}b
+#    The solver options for this are:
+#    'ksp_type': 'fgmres'
+#    We need fgmres here because gmres is used on the blocks.
+# b) Solve it with Picard iterations:
+#    Pu_{k+1} = (P - A)u_{k} + b
+#    The solver options for this are:
+#    'ksp_type': 'preonly'
+
+paradiag_parameters = {
+    'snes': {
+        'linesearch_type': 'basic',
+        'monitor': None,
+        'converged_reason': None,
+        'rtol': 1e-10,
+        'atol': 1e-12,
+        'stol': 1e-12,
+    },
+    'mat_type': 'matfree',
+    'ksp_type': 'preonly',
+    'ksp': {
+        'monitor': None,
+        'converged_reason': None,
+        'rtol': 1e-10,
+        'atol': 1e-12,
+        'stol': 1e-12,
+    },
+    'pc_type': 'python',
+    'pc_python_type': 'asQ.DiagFFTPC'
+}
+
+# We need to add a block solver parameters dictionary for each block.
+# Here they are all the same but they could be different.
+for i in range(window_length):
+    paradiag_parameters['diagfft_block_'+str(i)+'_'] = block_parameters
+
+
+# # # === --- Setup ParaDiag --- === # # #
+
+
+# Give everything to asQ to create the paradiag object.
+# the circ parameter determines where the alpha-circulant
+# approximation is introduced. None means only in the preconditioner.
+PD = asQ.paradiag(ensemble=ensemble,
+                  form_function=form_function,
+                  form_mass=form_mass,
+                  w0=w0, dt=dt, theta=args.theta,
+                  alpha=args.alpha, time_partition=time_partition, bcs=bcs,
+                  solver_parameters=paradiag_parameters,
+                  circ=None)
+
+
+# This is a callback which will be called before pdg solves each time-window
+# We can use this to make the output a bit easier to read
+def window_preproc(pdg, wndw):
+    PETSc.Sys.Print('')
+    PETSc.Sys.Print(f'### === --- Calculating time-window {wndw} --- === ###')
+    PETSc.Sys.Print('')
+
+
+qcheck = w0.copy(deepcopy=True)
+
+
+def Exact(u_exact):
+    q_err = fd.errornorm(u_exact, qcheck)
+    return q_err
+
+
+def window_postproc(pdg, wndw):
```

Comment:
I think that this is quite an old method of getting the errors from before we had a lot of the current QoL methods on `AllAtOnceSystem`. I think it can be done in a much cleaner way with `SharedArray`.

Changing this is possibly better done in another PR though, along with the shallow water examples where we do a similar thing.

Reply:
Please can you collect errors using the same method as in `heat/heat.py` with `SharedArray` instead of using seperate `errors` and `local_errors` arrays with `for_each_timestep`?